requests-for-comments/2022-11-02-data-exploration.md

```diff
@@ -0,0 +1,600 @@
+
+```json
+{
+  "type": "funnel",
+  "steps": [
+    {
+      "type": "events"
+    },
+    {
+      "type": "events",
+      "filters": {
+        "event": "$pageview",
+        "properties": [{ "key": "$browser", "equals": "Chrome" }]
+      }
+    }
+  ]
+}
+```
+
+```ts
+interface FunnelNode extends Node {
+  type: "funnel";
+  steps: EventsNode[];
+}
+```
+
+If we want people who dropped off at a step, broken down further, it'd be something like:
+
+```json
+{
+  "type": "persons",
+  "breakdown": ["person.account_type"],
+  "from": {
+    "type": "funnel-dropoff",
+    "step": "2",
+    "from": {
+      "type": "funnels",
+      "steps": [
+        {
+          "type": "events"
+        },
+        {
+          "type": "events",
+          "filters": {
+            "event": "$pageview",
+            "properties": [
+              {
+                "key": "$browser",
+                "equals": "Chrome"
+              }
+            ]
+          }
+        }
+      ]
+    }
+  }
+}
+```
+
+```ts
+interface PersonsNode extends Node {
+  type: "persons";
+  breakdown?: string[];
+  from?: FunnelDropoffNode | EventsNode;
+}
+
+interface FunnelDropoffNode extends Node {
+  type: "funnel-dropoff";
+  step: number;
+  from: FunnelNode;
+}
+```
+
+I'm not sure how many levels of abstractions this requires. It'll depend on how this maps to actual SQL.
+
+You could get correlation analysis for a funnel in a similar way:
+
+```json
+{
+  "type": "correlation",
+  "correlation": {
+    "type": "properties",
+    "exclude_names": [
+      "$initial_geoip_postal_code",
+      "$initial_geoip_latitude",
+      "$initial_geoip_longitude"
+    ],
+    "names": ["$all"]
+  },
+  "from": {
+    "type": "funnels",
+    "steps": [
+      {
+        "type": "events"
+      },
+      {
+        "type": "events",
+        "filters": {
+          "event": "$pageview",
+          "properties": [{ "key": "$browser", "equals": "Chrome" }]
+        }
+      }
+    ]
+  }
+}
+```
+
+```ts
+interface CorrelationNode extends Node {
+  type: "correlation";
+  correlation?: CorrelationOptions;
+  from?: FunnelNode;
+}
+```
+
+## Backend
+
+TODO: what is needed in the backend to make this possible?
+
+## Frontend
+
+### Iteration 1
+
+The "data exploration" view will basically be one big query editor. The first steps will look similar to the existing [MVP](https://github.com/PostHog/posthog/pull/11981) and expose a big table. 
+
+- Two select boxes and a large table below. 
+- You can select "events", apply "matching filters" and see a list of rows
+- You can switch "events" to "persons" and get a different list.
+- Switching to "recordings" might open a different looking interface (recordings might come later)
+
+The goal is to replace the existing events and persons tables, including the persons table in insight modals.
+
+These person modals will now have an "Expand" button that will open them in the "data explorer" view.
```

Comment:
I feel like already at this stage this waves way a lot of UI complexity with the "Date Explorer" view. This UX is the part that's unclear to me.

Reply:
Will expand on this.